### ISSUE TITLE
Discussion: Integrating directly into Godot

### ISSUE THREAD

**Issue by RadiantUwU** [godot-rust/gdext#782]
I think this would prove very useful for more Rust-Godot interoperability.
Making an engine module / plugin to allow users to directly use rust on Godot as a programming language interfacing with [ScriptLanguageExtension](https://docs.godotengine.org/en/latest/classes/class_scriptlanguageextension.html) and [ScriptExtension](https://docs.godotengine.org/en/latest/classes/class_scriptextension.html#class-scriptextension) would allow users to directly use Rust inside Godot Engine for their games providing high performance.

## Building

Similar to C#, there will be a build button to build all scripts. The lib.rs is automatically generated by Godot, including a loader and all rust scripts generated provided under modules. You may also have crate management to include other crates, but that should be done later.

## Module or GDExtension?

There are a few advantages to making an engine module over a GDExtension, for example:
- Being allowed to create new classes (RustScript, RustScriptLanguage(unexposed))
- Having more access to the engine's internals

In conclusion, we can use the script alongside GDExtension bindings to be able to freely run rust inside Godot for making it easier for the user to make games directly in Rust inside the game engine itself.

**Comment by Ughuuu** [comment]
Hi,

I created something similar to this using gdext you can look at, it provides syntax highlight and builds the rust code on save(there is an issue as it is called for every file, and had some issue with path where it finds Cargo.toml file).
https://github.com/Ughuuu/gdrust-syntax
You can use it as a starting point if you want.

**Build**

My thought was to build everything on script save, same as GDScript

**Module or GDExtension?**

GDExtension can do also script and script language, same as module.

https://docs.godotengine.org/en/stable/classes/class_scriptextension.html
https://docs.godotengine.org/en/stable/classes/class_scriptlanguageextension.html

You can see in my repo I did that.

To your last point, I see this more as a syntax highlighter and a build on save tool, rather than limiting too much. It would also not be in the scope of this repo, as gdext sets to create Rust bindings for Godot, not an extension with the said bindings that does extra work. So it would definitely be another repository of it's own.

**Comment by Bromeon** [comment]
Hello, and thanks for the proposal :slightly_smiling_face: 

I think however that module development is out of scope for the godot-rust project. We focus on integrating Rust via GDExtension API, which has several core advantages:

1. The API is very actively maintained and has its own dedicated Godot team. As a result, it has evolved significantly, and the interface is mostly a C header fine which is nowadays well-documented.
2. GDExtension and is used by languages from completely different backgrounds (C++, Rust, Go, Swift, Kotlin...). The module experience is very C++ focused on the other hand. Integrating a new language needs a custom bridging technology -- while GDExtension just needs interfacing with C, which every decent language supports out of the box.
3. Developing a module is generally harder, as it requires multiple toolchains. If I am a Rust developer, with gdext I only need to install Rust via rustup, which is trivial, and I can immediately start hacking. With modules, I additionally need Python, scons and Godot just to be able to compile the engine.
4. There is no compatibility story for modules, and no stable long-term interfaces. If I develop an extension with Godot 4.1, I can still use it in Godot 4.3. godot-rust even makes this [extra simple](https://godot-rust.github.io/book/toolchain/compatibility.html). For modules, I need to manually update code and it's very hard (if possible at all) to stay simultaneously compatible with multiple Godot versions. This in turn makes building an ecosystem much harder, because everything has to work with latest at all times, however in reality different maintainers have different pace.
5. Modules require a customized engine version, whereas extension can be distributed much more easily on top. If I have an existing setup for a Godot version (e.g. a team agreed to use 4.2.2), then extensions can be added non-intrusively, I can just ship them via Git. For modules, I need change everyone's version first. Modules can also not be enabled/disabled easily (without recompilation).

On the other hand, I don't think that the points you mention are big drawbacks.

> Being allowed to create new classes (RustScript, RustScriptLanguage(unexposed))

You can define your own classes in GDExtension (that's a main point of it). GDExtension also supports custom scripts, we have [a dedicated Rust API `obj::script`](https://godot-rust.github.io/docs/gdext/master/godot/obj/script/index.html) for it.

> Having more access to the engine's internals

While true, most projects don't need this and the existing GDExtension API surface is huge. The benefit of using official interfaces is that they remain stable and aren't broken in the next internal refactoring. If something internal is popular enough that extensions would benefit it, it's also possible to promote it to an official API in Godot.

**Comment by RadiantUwU** [comment]
Hello! Are we allowed to use the GDExtension glue from this repository to include into the new engine module?

**Comment by Bromeon** [comment]
As long as you respect the license, yes 🙂 i.e. code needs to stay public under MPL-2.0 and the copyright header must not be changed.

**Comment by RadiantUwU** [comment]
Im planning to use MIT license and to put it into Godot itself.

**Comment by Bromeon** [comment]
Then I'm afraid you can't use the original godot-rust code -- it needs to stay MPL-2.0. This is btw. not something I decide alone, every contributor has submitted as MPL-2.0 under the assumption that it stays open.

You can however see if there's a way to write your own glue in MIT and then depend on godot-rust.

**Comment by RadiantUwU** [comment]
I'm thinking if i can put it under the thirdparty folder
Also
> You may create and distribute a Larger Work under terms of Your choice, provided that You also comply with the requirements of this License for the Covered Software. If the Larger Work is a combination of Covered Software with a work governed by one or more Secondary Licenses, and the Covered Software is not Incompatible With Secondary Licenses, this License permits You to additionally distribute such Covered Software under the terms of such Secondary License(s), so that the recipient of the Larger Work may, at their option, further distribute the Covered Software under the terms of either this License or such Secondary License(s).

**Comment by Bromeon** [comment]
Yes, you can redistribute gdext as long as you keep its license.

Maybe as a general question: do you have concrete use cases where someone would want to use Rust, but _not_ GDExtension? The initial development effort likely won't be enough; such an integration needs to be maintained for possibly years, which is a big commitment.

**Comment by RadiantUwU** [comment]
> Yes, you can redistribute gdext as long as you keep its license.
> 
> Maybe as a general question: do you have concrete use cases where someone would want to use Rust, but _not_ GDExtension? The initial development effort likely won't be enough; such an integration needs to be maintained for possibly years, which is a big commitment.

The only reason is direct integration, just like C#.
Except C# requires Visual studio, for Godot, we won't need that.

**Comment by Bromeon** [comment]
You mean to write Rust code inside the Godot editor? That should work using GDExtension. Check @Ughuuu's [message](https://github.com/godot-rust/gdext/issues/782#issuecomment-2192037566) above, maybe you guys can collaborate on something :slightly_smiling_face:

**Comment by Ughuuu** [comment]
My feel is that godot wants to move more towards GDExtension instead of a fully packed engine. When downloading the engine you would get a list of some addons, some of which would be official ones. And you can install the most used ones.

I dont think there are things you cannot do as an addon, from what you wanted to do.

I do think it would be much more easy for the user to have to not install any addon and have it preinstalled. Another alternative to this would be global addons, which you install once and it works for whole godot.

If you do go the route of making it a module, though, you have to be very certain it will get into the engine, because otherwise, as @Bromeon said, people will need a custom engine build or a separate download to get this engine.

I know that currently the c# one is a module, but even then there are 2 game engine builds, its not on by default for all people. And the thought on that is to be moved to a gdextension as well.
From https://github.com/godotengine/godot-proposals/issues/7895 : "Ideally, we want to move Godot to use the GDExtension API instead of the
current custom glue glue"

**Comment by Bromeon** [comment]
> If you do go the route of making it a module, though, you have to be very certain it will get into the engine, because otherwise, as @Bromeon said, people will need a custom engine build or a separate download to get this engine.

To be realistic here, Rust is **very unlikely** to become integrated into core Godot, for a variety of reasons:
- Most Godot users don't benefit from it, but it would increase binary size for everyone.
- The toolchain and build system becomes more complex.
- The maintenance would end up with core Godot maintainers, but Rust isn't an area of expertise within the team, and the resources are needed elsewhere.
- We don't have the same stability guarantees as the rest of the engine.
- We have a different update/release schedule, meaning people couldn't get newer godot-rust features for a long time.
- There is, today, a perfectly working solution that is super easy to install: GDExtension. Now with the crates.io release, it's even easier.

Also, the [entire reason](https://docs.godotengine.org/en/stable/tutorials/scripting/gdextension/what_is_gdextension.html) why GDExtension was created is to allow third-party projects (like godot-rust) to be developed independently from the core engine, thus not being bottlenecked on core contributors.

Don't get me wrong, it's good to explore different directions, but it's a pity if people spend time on approaches that have almost no chance of happening, and it will cause frustration :confused: also, "Rust must be a core module" isn't something a user has ever complained about, so I'm not sure we're fixing any user needs.

Instead, we should identify what exactly is missing for certain users (e.g. the editor UI you started building), and spend efforts on those :slightly_smiling_face:

**Comment by RadiantUwU** [comment]
Since no rust files require to be compiled, as all of them are just included as-is into the project directory, it won't really add another dependency to the build system.

**Comment by Bromeon** [comment]
> C# has been added into the game engine, why would Rust be unable to get added?
> Godot provides builds with and without C# support

C# has been part of Godot for many years, and it enjoys much greater popularity in the gamedev ecosystem. Also, speaking of which, there are discussions to move C# towards GDExtension...

---

> This could be said for literally anything. RenderingDevice being exposed, C#, Line2Ds, Expression class. So many things, yet they come in since for the rare amount of people that do need it, they would feel it missing if it didnt exist.

This is precisely the point. Using this argument, you can justify any feature because it's always useful for some people. A big differentiator here is "can it be done as an extension", which is true for Rust. And there needs to be a balance between the number of users benefiting from a feature, vs. the overhead it creates. If Rust, why not also Swift, Kotlin, Go, and others?

---

> This is not for the core maintainers. This is for the gdextension team.

I happen to be part of that team 😉 but if it's built-in as a module into the engine, it may touch other areas, such as editor, buildsystem, core, etc.

---


Either way, I've voiced my concerns. To check the feasibility, this discussion should be brought up to https://github.com/godotengine/godot-proposals rather than this repo. But be warned, there have been similar discussions in the past 🙂

**Comment by Bromeon** [comment]
Closing, as I haven't received answers to most of my arguments in this discussion, and this should be brought up to Godot. In the meantime, feel free to reach out in case you need help with gdext!
